### ISSUE TITLE
Address long runtimes of test suite

### ISSUE THREAD

**Issue by lbianchi-lbl** [watertap-org/watertap#678]
- [x] Configure `concurrency` for PR GHA workflow to avoid "CI pileups" (#1115)
- [ ] Move repeated tests that exercise identical parts of the code (i.e. parameter sweeps) to dedicated pytest marker and/or less-frequent workflows
- [x] Make pytest jobs dependent on formatting check so that jobs don't run needlessly if unformatted code is pushed to an open PR

**Comment by lbianchi-lbl** [comment]
Summarizing the discussion with @bknueven during the dev call:

- The first step would be to identify which tests contribute the most to the runtime and are narrow-scoped (i.e. the causes that could make them fail are limited)
  - Without knowing the details, this could apply to tests that run parameter sweeps (i.e. they exercise the same parts of the Python API, but produce different results)
- Given the amount of PRs that are active at any given time, the approach used for IDAES (integration tests run once every PR instead than for every commit) probably would not give us much in terms of reduced runtime (since merging PR "A"  would require the tests to run again before merging PR "B")
- Instead, the current idea would be the following:
  - By default, long-running/integration tests are run on a schedule (e.g. nightly), but instead of running only for the latest commit on `main`, they run on every commit that has been merged since the latest run
    - In other words, this is similar to running `git bisect` over however many commits were merged during the previous 24 h
    - This ensures that every commit is tested (so that we can pinpoint where any breaking change comes from), but without causing a bottleneck when many PRs need to be merged (since the jobs run asynchronously outside of "PR rush hour")
  - As suggested by @bknueven: In addition to the scheduled runs, we should make it possible to optionally trigger the same tests for a specific PR (e.g. by adding a special label, which is how IDAES does it)
     - This covers the case when a developer knows that the PR they're working on could be "risky" (i.e. its changes would cause the tests to fail), they have the option to run the tests *before* merging (and thereby avoiding to merge, wait for the tests to run, reopen another PR to fix, and so on)

**Comment by lbianchi-lbl** [comment]
- The main conceptual obstacle on my side is figuring out the details of how to trigger the "on-demand" integration tests for PRs
- @bknueven sensibly suggests to use a sentinel label (i.e., the trigger is the `issue.labeled` event), similarly to how this functionality is implemented in IDAES/idaes-pse
- I think this is a viable solution, however from the IDAES/idaes-pse experience there are two main issues still to solve:
  - The requirement to manually add a label is not easy to discover and/or easy to forget for developers
  - As a way to address this, it should now be possible to have other workflows assign the label automatically from other triggering events. However, this in turn causes race conditions and other hard-to-understand behaviors that cause e.g. running jobs to be cancelled or invalidated in ways that are equally confusing for developers (if not even more so)
- In summary, the `issue.labeled` trigger could be a reasonable intermediate solution

**Comment by lbianchi-lbl** [comment]
Moving this to the next release after #1156 resolved some of the intended goals.

**Comment by lbianchi-lbl** [comment]
Closing this as the low-hanging fruits have been completed, and the remaining item is looking a bit too far up to be done anytime soon.
